lib/jsonapi/utils/underscore.ex

```diff
@@ -56,14 +60,44 @@ defmodule JSONAPI.Utils.Underscore do
   end
 
   def underscore(value) when is_map(value) do
-    Enum.into(value, %{}, &underscore/1)
+    Enum.into(value, %{}, &underscore(&1))
   end
 
   def underscore({key, value}) do
-    if is_map(value) do
-      {underscore(key), underscore(value)}
+    if underscore?(key) do
+      if is_map(value) do
+        {underscore(key), underscore(value)}
+      else
+        {underscore(key), value}
+      end
     else
       {underscore(key), value}
     end
   end
+
+  defp underscore?(key) do
+    config = Application.get_env(:jsonapi, :underscore_to_dash)
+    config_specifies_underscore?(config, key)
+  end
+
+  defp config_specifies_underscore?(true, _), do: true
+  defp config_specifies_underscore?(false, _), do: false
+
+  defp config_specifies_underscore?(config, key) when is_list(config) do
+    if Keyword.has_key?(config, :only) do
+      config_specifies?(config, :only, key)
+    else
+      !config_specifies?(config, :except, key)
+    end
+  end
+
+  defp config_specifies?(config, only_or_except, key) when is_list(config) do
```

Comment:
Do we need this this `is_list/1` guard?  There's no other `config_specifies?/3` functions that could be called.  If we intend to handle non-list cases, we should define that function.

I'm not sure about the `case` and `is_list/1` inside either 🤔.  If a user supplies an invalid configuration (a non-list) such that we need `is_list/1` _should_ we proceed quietly?  That seems like it would not be expected by the user.

I'm wondering if we should just simplify this and in the event they pass invalid configuration it'll raise an error:

```elixir
defp config_specifies?(config, only_or_except, key) do
  key in Keyword.get(config, only_or_except, [])
end
```

Reply:
yes!